### PR TITLE
WoW end2end interactive.

### LINKED ISSUE #1882: Talk interactively to the generative WoW model
Is there a way I can talk to the **generative** Wizard of Wikipedia model interactively? There's a command provided for the retrieval model in the [project README](https://parl.ai/projects/wizard_of_wikipedia/), but none for the generative model.

I tried the following:

`python examples/interactive.py -mf models:wizard_of_wikipedia/end2end_generator/model -t wizard_of_wikipedia`

but that gave the following error:

```
Please choose one of the following topics by typing A, B, C, ..., etc. :

A: Art
B: Graphic design
C: Tokyo
D: Author
E: [NO TOPIC]

Enter Your Message: C
[ Your chosen topic is: Tokyo ]
Enter Your Message: I've never been to Japan, have you?
Traceback (most recent call last):
  File "examples/interactive.py", line 18, in <module>
    interactive(opt, print_parser=parser)
  File "/juice/scr/scr101/scr/abisee/ParlAI/parlai/scripts/interactive.py", line 79, in interactive
    world.parley()
  File "/juice/scr/scr101/scr/abisee/ParlAI/parlai/tasks/wizard_of_wikipedia/worlds.py", line 93, in parley
    self.acts[1] = self.model_agent.act()
  File "/juice/scr/scr101/scr/abisee/ParlAI/parlai/core/torch_agent.py", line 1662, in act
    return self.batch_act([self.observation])[0]
  File "/juice/scr/scr101/scr/abisee/ParlAI/parlai/core/torch_agent.py", line 1683, in batch_act
    batch = self.batchify(observations)
  File "/juice/scr/scr101/scr/abisee/ParlAI/projects/wizard_of_wikipedia/generator/agents.py", line 211, in batchify
    obs_know = self._parse_knowledge(obs)
  File "/juice/scr/scr101/scr/abisee/ParlAI/projects/wizard_of_wikipedia/generator/agents.py", line 172, in _parse_knowledge
    obs['title'], TOKEN_KNOWLEDGE, obs['checked_sentence']
KeyError: 'title'
```

Thanks!

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Hacks together an interactive mode for the End2End wizard of wikipedia agent. Unlike the ones used in the actual mechanical turk experiments, this one doesn't have any protection preventing it from using the same knowledge twice, so it's a bit more repetitive than it should be. Nonetheless, it lets one play.

Fixes #1882.

**Testing steps**
`python examples/interactive.py -m projects:wizard_of_wikipedia:interactive_end2end -t wizard_of_wikipedia`

**Logs**
```
Please choose one of the following topics by typing A, B, C, ..., etc. :

A: Army
B: World Wide Web
C: Broccoli
D: London Marathon
E: [NO TOPIC]

Enter Your Message: D
[ Your chosen topic is: London Marathon ]
[WizardEnd2EndInteractiveAgent]: i ' ve always wanted to go to the london trail .
Enter Your Message: do you like to run much?
[WizardEnd2EndInteractiveAgent]: no , but i would love to go to london . it ' s a long time ago .
Enter Your Message: what spot do you want to see in london the most?
[WizardEnd2EndInteractiveAgent]: i ' m not sure , but i do know that it has been around a long time .
Enter Your Message: that's true. i'm most interested in seeing big ben
[WizardEnd2EndInteractiveAgent]: yes , it has been around a long time .
```

### CODE REVIEW

File: projects/wizard_of_wikipedia/generator/agents.py [hunk 1]
```diff
@@ -202,7 +209,7 @@ def batchify(self, obs_batch):
         """
         batch = super().batchify(obs_batch)
         reordered_observations = [obs_batch[i] for i in batch.valid_indices]
-        is_training = not ('eval_labels' in reordered_observations[0])
+        is_training = 'labels' in reordered_observations[0]
```
Comment (emilydinan) [inline projects/wizard_of_wikipedia/generator/agents.py]: this gets me every time lol

File: projects/wizard_of_wikipedia/interactive_end2end/interactive_end2end.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+from parlai.core.agents import Agent, create_agent, create_agent_from_shared
+from projects.wizard_of_wikipedia.generator.agents import EndToEndAgent
+
+from parlai.tasks.wizard_of_wikipedia.agents import TOKEN_KNOWLEDGE
+
+import json
+import os
+
+
+class InteractiveEnd2endAgent(Agent):
```
Comment (emilydinan) [inline projects/wizard_of_wikipedia/interactive_end2end/interactive_end2end.py]: can this inherit from the retrieval one or nah?

i can open an issue to move the tfidf stuff into the world

Comment (stephenroller) [reply projects/wizard_of_wikipedia/interactive_end2end/interactive_end2end.py]: That'd be great.

So some of the issues:

- We need a way for the agent to mark knowledge as not to be picked again. When the knowledge was handled separately by the two stage agent, that was easy enough
- The generative model expects that KNOWLEDGE token to be separating them, along with a title.

File: projects/wizard_of_wikipedia/interactive_end2end/interactive_end2end.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            'model_file': opt['retriever_model_file'],
+            'remove_title': False,
+            'datapath': opt['datapath'],
+            'override': {'remove_title': False},
+        }
+        self.retriever = create_agent(retriever_opt)
+
+        self._set_up_sent_tok()
+        wiki_map_path = os.path.join(self.model_path, 'chosen_topic_to_passage.json')
+        self.wiki_map = json.load(open(wiki_map_path, 'r'))
```
Comment (klshuster) [inline projects/wizard_of_wikipedia/interactive_end2end/interactive_end2end.py]: note that you leave this file open; perhaps use the `with open(...` syntax?

**Pull request merged**